File: Rust/src/common/rangevector.rs
use crate::util::check_argument;
use crate::types::Result;

/**
 * A RangeVector is used when we want to track a quantity and its upper and
 * lower bounds
 */
#[repr(C)]
#[derive(Clone)]
pub struct RangeVector<T> {
    pub values: Vec<T>,
    pub upper: Vec<T>,
    pub lower: Vec<T>
}

impl RangeVector<f32> {
    pub fn new(dimensions: usize) -> Self {
        RangeVector {
            values: vec![0.0; dimensions],
            upper: vec![0.0; dimensions],
            lower: vec![0.0; dimensions]
        }
    }
}

impl RangeVector<f64> {
    pub fn new(dimensions: usize) -> Self {
        RangeVector {
            values: vec![0.0; dimensions],
            upper: vec![0.0; dimensions],
            lower: vec![0.0; dimensions]
        }
    }
}


impl<T: PartialOrd + Clone + Copy  + std::ops::AddAssign + std::ops::MulAssign> RangeVector<T> {
    pub fn from(values : Vec<T>) -> Self {
        RangeVector{
            values : values.clone(),
            upper : values.clone(),
            lower : values.clone()
        }
    }

    pub fn create(values: &[T], upper: &[T], lower:&[T]) -> Result<Self> {
        check_argument(values.len() == upper.len() && upper.len() == lower.len(), " incorrect lengths")?;
        for i in 0..values.len() {
            check_argument(values[i] <= upper[i], " incorrect upper bound")?;
            check_argument(lower[i] <= values [i], "incorrect lower bounds")?;
        }
        Ok(RangeVector{
            values :Vec::from(values),
            upper : Vec::from(upper),
            lower : Vec::from(lower)
        })
    }

    pub fn shift(&mut self, i:usize, shift: T) {
        self.values[i] += shift;
        self.upper[i] += shift;
        self.lower[i] += shift;
        // managing precision explicitly
        if self.upper[i] < self.values[i] {
            self.upper[i] = self.values[i];
        }
        if self.lower[i] > self.values[i] {
            self.lower[i] = self.values[i];
        }
    }

    pub fn cascaded_add(&mut self, base: &[T]) -> Result<()>{
        check_argument(base.len() >0 , "must be of positive length")?;
        let horizon = self.values.len()/base.len();
        check_argument(horizon * base.len() == self.values.len(), " incorrect function call")?;
        for j in 0..base.len() {
            self.shift(j,base[j]);
        }
        for i in 1..horizon {
            for j in 0..base.len() {
                self.shift(i * base.len() + j, self.values[(i-1)*base.len() + j]);
            }
        }
        Ok(())
    }

    pub fn scale(&mut self, i:usize, scale: T) {
        self.values[i] *= scale;
        self.upper[i] *= scale;
        self.lower[i] *= scale;
        // managing precision explicitly
        if self.upper[i] < self.values[i] {
            self.upper[i] = self.values[i];
        }
        if self.lower[i] > self.values[i] {
            self.lower[i] = self.values[i];
        }
    }
}

File: Java/core/src/main/java/com/amazon/randomcutforest/returntypes/RangeVector.java
/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

package com.amazon.randomcutforest.returntypes;

import static com.amazon.randomcutforest.CommonUtils.checkArgument;
import static java.lang.Math.max;
import static java.lang.Math.min;

import java.util.Arrays;

/**
 * A RangeVector is used when we want to track a quantity and its upper and
 * lower bounds
 */
public class RangeVector {

    public final float[] values;

    /**
     * An array of values corresponding to the upper ranges in each dimension.
     */
    public final float[] upper;
    /**
     * An array of values corresponding to the lower ranges in each dimension
     */
    public final float[] lower;

    public RangeVector(int dimensions) {
        checkArgument(dimensions > 0, "dimensions must be greater than 0");
        values = new float[dimensions];
        upper = new float[dimensions];
        lower = new float[dimensions];
    }

    /**
     * Construct a new RangeVector with the given number of spatial dimensions.
     * 
     * @param values the values being estimated in a range
     * @param upper  the higher values of the ranges
     * @param lower  the lower values in the ranges
     */
    public RangeVector(float[] values, float[] upper, float[] lower) {
        checkArgument(values.length > 0, " dimensions must be > 0");
        checkArgument(values.length == upper.length && upper.length == lower.length, "dimensions must be equal");
        for (int i = 0; i < values.length; i++) {
            checkArgument(upper[i] >= values[i] && values[i] >= lower[i], "incorrect semantics");
        }
        this.values = Arrays.copyOf(values, values.length);
        this.upper = Arrays.copyOf(upper, upper.length);
        this.lower = Arrays.copyOf(lower, lower.length);
    }

    public RangeVector(float[] values) {
        checkArgument(values.length > 0, "dimensions must be > 0 ");
        this.values = Arrays.copyOf(values, values.length);
        this.upper = Arrays.copyOf(values, values.length);
        this.lower = Arrays.copyOf(values, values.length);
    }

    /**
     * Create a deep copy of the base RangeVector.
     *
     * @param base The RangeVector to copy.
     */
    public RangeVector(RangeVector base) {
        int dimensions = base.values.length;
        this.values = Arrays.copyOf(base.values, dimensions);
        this.upper = Arrays.copyOf(base.upper, dimensions);
        this.lower = Arrays.copyOf(base.lower, dimensions);
    }

    public void shift(int i, float shift) {
        checkArgument(i >= 0 && i < values.length, "incorrect index");
        values[i] += shift;
        // managing precision
        upper[i] = max(values[i], upper[i] + shift);
        lower[i] = min(values[i], lower[i] + shift);
    }

    public void scale(int i, float weight) {
        checkArgument(i >= 0 && i < values.length, "incorrect index");
        checkArgument(weight > 0, " negative weight not permitted");
        values[i] = values[i] * weight;
        // managing precision
        upper[i] = max(upper[i] * weight, values[i]);
        lower[i] = min(lower[i] * weight, values[i]);
    }

}
